### ISSUE TITLE
Add fractional support to PostgreSQLIntervalType

### ISSUE THREAD

**Issue by lion7** [vladmihalcea/hibernate-types#307]
Hi @vladmihalcea, thanks for creating this great library! 😄 
I was wondering if there is a reason why intervals are being rounded to whole seconds?
Currently the precision of a Java `Duration` is up to nanoseconds and the precision of a PG `interval` is up to microseconds.
I'd like to create a PR which rounds Java duration to microseconds instead of whole seconds.
Any objections against such a change? 

The rounding happens here:

https://github.com/vladmihalcea/hibernate-types/blob/dd44df9fc418537cd5f19221b727c01367fd79a1/hibernate-types-52/src/main/java/com/vladmihalcea/hibernate/type/interval/PostgreSQLIntervalType.java#L47

**Comment by vladmihalcea** [comment]
I remember the PostgreSQL `Interval` column type can store up to seconds, as explained in [this article](https://www.2ndquadrant.com/en/blog/know-what-time-it-is/).

If you think microseconds are supported, then please provide a Pull Request that adds support for it and with a test case that shows that we can save the microseconds in the DB and read it afterward.

Looking forward to reviewing it.

**Comment by lion7** [comment]
It is actually mentioned in the article that you linked, be it very subtle:

> This list includes [(p)] which is, for example (3). This means that the type has precision 3 for milliseconds in the value. ‘p’ can be 0-6, but the type must include seconds.

PR coming soon 😄

**Comment by lion7** [comment]
By the way, is there a specific reason why the postgresql version is a bit old at version `9.4-1202-jdbc41`?
The latest versions for the `9.4.x` branch is `9.4-1212` and the absolute latest version is `42.2.19` according to https://mvnrepository.com/artifact/org.postgresql/postgresql.

**Comment by lion7** [comment]
I decided to copy some of the logic from the newer postgresql driver instead of updating the version.
Basically just [converting to micros](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/util/PGInterval.java#L362) and [converting from micros](https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/main/java/org/postgresql/util/PGInterval.java#L380).
The PR is here: https://github.com/vladmihalcea/hibernate-types/pull/308

There is some more info about the fraction and its resolution here: https://www.postgresql.org/docs/11/datatype-datetime.html

**Comment by lion7** [comment]
As a final note: if you decide to upgrade the postgresql version to the `42.2.x` branch then you can basically just use built-in conversion directly using `getWholeSeconds()` and `getMicroSeconds()`:
```java
        final int days = interval.getDays();
        final int hours = interval.getHours();
        final int minutes = interval.getMinutes();
        final int seconds = interval.getWholeSeconds();
        final int micros = interval.getMicroSeconds();

        return Duration.ofDays(days)
                .plus(hours, ChronoUnit.HOURS)
                .plus(minutes, ChronoUnit.MINUTES)
                .plus(seconds, ChronoUnit.SECONDS)
                .plus(micros, ChronoUnit.MICROS);
```

**Comment by vladmihalcea** [comment]
Thanks for the PR. The 52 branch can use Java 1.8, so it can use a newer Driver. The older versions must use Java 1.6, matching the associated Hibernate version.

**Comment by lion7** [comment]
Ah, that explains it a bit. There is also a newer version available that is compatible with Java 1.6: `42.2.19.jre6`. Altough I have no idea if it will work with older hibernate versions.

Back to the original question: do you want me to update the version as part of this PR? Of should I make a new one for the version upgrade?

**Comment by vladmihalcea** [comment]
Thanks, I merged it.
